crates/holochain_types/src/dht_op.rs

```diff
@@ -493,14 +493,18 @@ impl DhtOpLight {
                 Self::RegisterUpdatedElement(header_hash, entry_hash, basis.into())
             }
             DhtOpType::RegisterDeletedBy => {
-                Self::RegisterDeletedBy(header_hash.clone(), header_hash.into())
+                let basis = match header {
+                    Header::Delete(delete) => delete.deletes_address.clone(),
+                    _ => return Err(DhtOpError::OpHeaderMismatch(op_type, header.header_type())),
+                };
+                Self::RegisterDeletedBy(header_hash, basis.into())
             }
             DhtOpType::RegisterDeletedEntryHeader => {
-                let basis = header
-                    .entry_hash()
-                    .ok_or_else(|| DhtOpError::HeaderWithoutEntry(header.clone()))?
```

Comment:
@freesig presumably this was the source of our troubles, it was the other candidate I discovered more recently, and your fix makes a lot more sense to me than the code that was here before!

I struggled to think of how I would have produced a 'steps to reproduce' situation for this, since it was fairly nuanced and seemed, from our user testing, since it relied on gossip, to happen basically by chance or not. In the sense it seemed very non-deterministic, or pinning down what defined its determinism was challenging if not impossible without controlling more granularly the drivers of the gossip engine. 
Am I off, or can you imagine a simple-ish way to achieve steps to reproduce of something like this, from a level such as regular app usage (calling zome functions) via tryorama or sweettest?

Reply:
we were able to confirm in a loose way the behaviour that would reliably NOT cause the issue, which was joining a highly populated DHT that had a total of ZERO Delete headers in it, and joining worked well regardless of the scale. 

Once there Delete headers, it worked unreliably, but still working sometimes, or partially.

Reply:
Yeh this is a very hard thing to reliably reproduce from a high level. On the other hand I could see from your logs that it was a delete header and there is only so many places in the codebase that a delete header can produce this problem. Unfortunately that means it's hard to create a high level test to prove this works.